quic/s2n-quic-core/src/stream/state/recv.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+
+//= https://www.rfc-editor.org/rfc/rfc9000#section-3.2
+//#        o
+//#       | Recv STREAM / STREAM_DATA_BLOCKED / RESET_STREAM
+//#       | Create Bidirectional Stream (Sending)
+//#       | Recv MAX_STREAM_DATA / STOP_SENDING (Bidirectional)
+//#       | Create Higher-Numbered Stream
+//#       v
+//#   +-------+
+//#   | Recv  | Recv RESET_STREAM
+//#   |       |-----------------------.
+//#   +-------+                       |
+//#       |                           |
+//#       | Recv STREAM + FIN         |
+//#       v                           |
+//#   +-------+                       |
+//#   | Size  | Recv RESET_STREAM     |
+//#   | Known |---------------------->|
+//#   +-------+                       |
+//#       |                           |
+//#       | Recv All Data             |
+//#       v                           v
+//#   +-------+ Recv RESET_STREAM +-------+
+//#   | Data  |--- (optional) --->| Reset |
+//#   | Recvd |  Recv All Data    | Recvd |
+//#   +-------+<-- (optional) ----+-------+
+//#       |                           |
+//#       | App Read All Data         | App Read Reset
+//#       v                           v
+//#   +-------+                   +-------+
+//#   | Data  |                   | Reset |
+//#   | Read  |                   | Read  |
+//#   +-------+                   +-------+
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub enum Receiver {
+    #[default]
+    Recv,
+    SizeKnown,
+    DataRecvd,
+    DataRead,
+    ResetRecvd,
+    ResetRead,
+}
+
+impl Receiver {
+    is!(Recv, is_receiving);
+    is!(SizeKnown, is_size_known);
+    is!(DataRecvd, is_data_received);
+    is!(DataRead, is_data_read);
+    is!(ResetRecvd, is_reset_received);
+    is!(ResetRead, is_reset_read);
+    is!(DataRead | ResetRead, is_terminal);
```

Comment:
seems like something that would be useful at some point to implement as a `derive` macro on the Enum

Reply:
Yeah possibly. I'm sure there's one that already exists. The question is it worth the dependency 😄